File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        // use hashmap to get frequency of numbers
        Map<Integer, Integer> map = new HashMap<>();

        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        // naive approach: use min-heap, contantly remove and add to get highest 3 frequencies 
        PriorityQueue<Integer> q = new PriorityQueue<>((a, b) -> map.get(a) - map.get(b));
        for (int num : map.keySet()) {
            q.add(num);
            if (q.size() > k) {
                q.poll();
            }
        }

        int[] res = new int[k];
        for (int i = 0; i < k; i++) {
            res[i] = q.poll();
        }

        return res;
    }
}

File: Data Structures & Algorithms/maximum-product-subarray/submission-0.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        minn = 1
        maxx = 1
        res = nums[0]

        for num in nums:
            curr_max = maxx * num
            curr_min = minn * num
            minn = min(curr_min, curr_max, num)
            maxx = max(curr_max, curr_min, num)
            res = max(res, maxx)

        return res